godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Variant.kt

```diff
@@ -0,0 +1,585 @@
+            Vector2::class -> nil.asVector2()
+            Vector3::class -> nil.asVector3()
+            else -> throw UnsupportedOperationException("Unknown variant class ${T::class}")
+        } as T
+    }
+}
+
+
+//FAKE CONSTRUCTORS. Necessary because inline classes don't support secondary constructors yet.
+//NULL
+fun Variant() = Variant.buildAny(Godot.gdnative.godot_variant_new_nil)
+
+//PRIMITIVES
+fun Variant(from: Boolean) = Variant.buildAny(Godot.gdnative.godot_variant_new_bool, from)
+fun Variant(from: Int) = Variant.buildAny(Godot.gdnative.godot_variant_new_int, from.toLong())
+fun Variant(from: Long) = Variant.buildAny(Godot.gdnative.godot_variant_new_int, from)
+fun Variant(from: Float) = Variant.buildAny(Godot.gdnative.godot_variant_new_real, from.toDouble())
+fun Variant(from: Double) = Variant.buildAny(Godot.gdnative.godot_variant_new_real, from)
+fun Variant(from: String) = from.asGDString { Variant.buildAny(Godot.gdnative.godot_variant_new_string, it.ptr) }
+
+//OBJECT
+fun Variant(from: Object) = Variant.buildAny(Godot.gdnative.godot_variant_new_object, from.ptr)
+
+//CORE
+fun Variant(from: AABB) = Variant.buildAny(Godot.gdnative.godot_variant_new_aabb, from)
+fun Variant(from: Basis) = Variant.buildAny(Godot.gdnative.godot_variant_new_basis, from)
+fun Variant(from: Color) = Variant.buildAny(Godot.gdnative.godot_variant_new_color, from)
+fun Variant(from: NodePath) = Variant.buildAny(Godot.gdnative.godot_variant_new_node_path, from)
+fun Variant(from: Plane) = Variant.buildAny(Godot.gdnative.godot_variant_new_plane, from)
+fun Variant(from: Quat) = Variant.buildAny(Godot.gdnative.godot_variant_new_quat, from)
+fun Variant(from: RID) = Variant.buildAny(Godot.gdnative.godot_variant_new_rid, from)
+fun Variant(from: Vector2) = Variant.buildAny(Godot.gdnative.godot_variant_new_vector2, from)
+fun Variant(from: Vector3) = Variant.buildAny(Godot.gdnative.godot_variant_new_vector3, from)
+
+//CONTAINER CORE
+fun Variant(from: PoolByteArray) = Variant.buildAny(Godot.gdnative.godot_variant_new_pool_byte_array, from)
+fun Variant(from: PoolColorArray) = Variant.buildAny(Godot.gdnative.godot_variant_new_pool_color_array, from)
+fun Variant(from: PoolIntArray) = Variant.buildAny(Godot.gdnative.godot_variant_new_pool_int_array, from)
+fun Variant(from: PoolRealArray) = Variant.buildAny(Godot.gdnative.godot_variant_new_pool_real_array, from)
+fun Variant(from: PoolStringArray) = Variant.buildAny(Godot.gdnative.godot_variant_new_pool_string_array, from)
+fun Variant(from: PoolVector2Array) = Variant.buildAny(Godot.gdnative.godot_variant_new_pool_vector2_array, from)
+fun Variant(from: PoolVector3Array) = Variant.buildAny(Godot.gdnative.godot_variant_new_pool_vector3_array, from)
+fun <T> Variant(from: VariantArray<T>) = Variant.buildAny(Godot.gdnative.godot_variant_new_array, from)
+
+//Generic constructor to wrap objects unrelated to Godot into a null variant
+fun Variant(from : Any?) = Variant()
```

Comment:
I think this might cause some ambiguity errors, passing an `AABB` for example can match `fun Variant(from: AABB)` and `fun Variant(from: Any?)`

Reply:
Yeah, it does.
But I had to put this constructor here because there is a dependency chain in the library. This constructor was is necessary because the VariantArray placeholder is using it with 
`    fun T.toVariant() = Variant(this)`.
It's necessary to cast the genetic type of the VariantArray into a Variant

```
    fun add(index: Int, element: T) {
        callNative {
            checkNotNull(Godot.gdnative.godot_array_insert)(it, index, element.toVariant()._handle.ptr )
        }
    }
```

That's one of the problems I got dealing with a generic VariantArray.